### ISSUE TITLE
CircleCi is not fetching the last fixpoint submodule

### ISSUE THREAD

**Issue by nikivazou** [ucsd-progsys/liquidhaskell#1566]
E.g. FP is properly updated but develop is fetched by circle in the following commit: 

https://github.com/ucsd-progsys/liquidhaskell/tree/48f6f82b3bb93c09115fa3a0fe930e302add0a92

**Comment by nikivazou** [comment]
Reported in stack: https://github.com/commercialhaskell/stack/issues/5080

**Comment by adinapoli** [comment]
Hey @nikivazou! I have been looking at this issue and I hope I can say with a certain degree of certainty I have found the culprit.  To reproduce the issue, I have created my own minified Github project which uses a submodule, to mimic what `liquidhaskell` does here with `liquid-fixpoint`:

https://github.com/adinapoli/circle-ci-git-submodules

As you can see, the idea is simple: I depend on a separate [git project used as a submodule](https://github.com/adinapoli/circle-ci-mysubmodule) to provide a simple `exposed :: Int` function that I can use in my main project as part of a testsuite:

```haskell
module Main where

import Test.Tasty
import Test.Tasty.HUnit

import MySubmodule (exposed)

main :: IO ()
main = defaultMain tests

tests :: TestTree
tests = testGroup "Tests" [unitTests]

--
unitTests = testGroup "Unit tests"
  [ testCase "Exposed" $
      exposed `compare` 42 @?= EQ
  ]
``` 

This way, I can test what happens when I bump my submodule to increment the value of `exposed`, and still have CI report success or failure if the number matches what I'm expecting in the main module.

After that, I have configured CircleCI for `circle-ci-git-submodules` and then opened this [PR](https://github.com/adinapoli/circle-ci-git-submodules/pull/1), that essentially replicates the usual scenario where `liquid-fixpoint` is bumped as part of a particular branch within `liquidhaskell`. After that, I was successfully able to replicate the CI problem, as you can see for example as part of this build:

https://circleci.com/gh/adinapoli/circle-ci-git-submodules/12

This is essentially the same that was happening to you as part of https://github.com/ucsd-progsys/liquidhaskell/pull/1574 : Note how my pull request bumps the submodule from SHA `b9c45f2e346bc8d27b237be775646367b69c6d7b` to `72b393304d15de4571973d19a89968aca3cd8401`, (so the latter is really what should be checked out) but during the step "Submodules" CircleCI rather checks-out  `b9c45f2e346bc8d27b237be775646367b69c6d7b`, which triggers the test failure.

I have re-run the build using the SSH support and logged into the VM, and typed:

```
root@2278f22181e8:~/project# git log -p --submodule 
commit 43ba4a122d44fae40fe0667ae930e235ea6ae4e9 (HEAD, origin/adinapoli/bump-exposed)
Author: Alfredo Di Napoli <alfredo.dinapoli@gmail.com>
Date:   Mon Jan 13 14:34:14 2020 +0100

    Bump exposed

diff --git a/Main.hs b/Main.hs
index a4836d5..68dbe91 100644
--- a/Main.hs
+++ b/Main.hs
@@ -11,7 +11,8 @@ main = defaultMain tests
 tests :: TestTree
 tests = testGroup "Tests" [unitTests]
 
+--
 unitTests = testGroup "Unit tests"
   [ testCase "Exposed" $
-      exposed `compare` 42 @?= EQ
+      exposed `compare` 43 @?= EQ
   ]
Submodule circle-ci-mysubmodule b9c45f2..72b3933:
  > Exposed = 43
```

So the git index tree was _aware_ of the submodule changes, but yet `git submodule update` was pulling the wrong checkout. So my suspect was that something was misbehaving _at the clone stage_, so that by the time we ran `git submodule update`, the checked-out submodule was already compromised.

What I did then was to try the next obvious step, i.e. trying to see if CircleCI's `checkout` command was to blame. After all [the internet](https://discuss.circleci.com/t/checkout-step-failed-with-submodules/29958) says that:

> The checkout actually runs a fairly involved shell script - it is not just a git clone. Since you have submodules, it may be worth skipping that step and writing your own. I believe there are enough environment variables available for you to see what branch/hash to checkout to.

So my next attempt was to try  replacing `checkout` with a custom command, but luckily I didn't need to go there, as there was one more thing I wanted to try first. In particular, if you look at the "Checkout code" step, you will notice a seemingly-innocuous warning coming from CircleCI:

> Either git or ssh (required by git to clone through SSH) is not installed in the image. Falling back to CircleCI's native git client but the behavior may be different from official git. If this is an issue, please use an image that has official git and ssh installed.

I thought that I was better be safe than sorry, and tried to install both `git` and `ssh` manually, and sure enough installing `ssh` manually did the trick:

https://circleci.com/gh/adinapoli/circle-ci-git-submodules/17

All it took was to partially modify `.circleci/config.yml` and add:

```
    steps:
      - add_ssh_keys
      - run: apt-get update && apt-get install -y z3 git ssh
      - checkout
```

While installing `git` is not necessary as it's already installed and doing the above will merely update it, what fixed the issue was installing `ssh`. Indeed after I have done that, the correct SHA is checked-out:

```
git submodule init
git submodule update
Submodule 'circle-ci-mysubmodule' (https://github.com/adinapoli/circle-ci-mysubmodule.git) registered for path 'circle-ci-mysubmodule'
Cloning into '/root/project/circle-ci-mysubmodule'...
Submodule path 'circle-ci-mysubmodule': checked out '72b393304d15de4571973d19a89968aca3cd8401'
```

## In summary

Something about how CircleCI handles submodules and SSH keys is fishy and the CircleCI people might be able to tell us exactly what's happening, but installing `ssh` manually is a simple and innocuous-enough fix to disregard more sophisticated investigations.

**Comment by nikivazou** [comment]
Thanks a lot @adinapoli ! Makes sense.

**Comment by nikivazou** [comment]
BTW, @adinapoli I run into this issue when I updated lts so I created this issue https://github.com/commercialhaskell/stack/issues/5080
Do you think it is irrelevant with lts upgrade and we should close the issue on commercialhaskell?

**Comment by adinapoli** [comment]
Hey Niki!

I think we should indeed close the issue -- I suspect the fact the problem manifested while you were upgrading snapshots was a red herring here. We can always re-open in the future in case we'll observe a stronger correlation 😉
